distribution/bwc/build.gradle

```diff
@@ -89,40 +89,69 @@ bwcVersions.forPreviousUnreleased { BwcVersions.UnreleasedVersionInfo unreleased
+        new ByteArrayOutputStream().withStream { os ->
+            ExecResult result = project.exec { spec ->
+                workingDir = checkoutDir
+                standardOutput os
+                action.execute(spec)
+            }
+            result.assertNormalExitValue()
+            return os.toString().trim()
         }
     }
-
-    File buildMetadataFile = project.file("build/${project.name}/build_metadata")
-    task writeBuildMetadata(type: LoggedExec) {
-        dependsOn checkoutBwcBranch
-        workingDir = checkoutDir
-        commandLine = ['git', 'rev-parse', 'HEAD']
-        ignoreExitValue = true
-        ByteArrayOutputStream output = new ByteArrayOutputStream()
-        standardOutput = output
+    task checkoutBwcBranch() {
+        dependsOn fetchLatest
         doLast {
-            if (execResult.exitValue != 0) {
-                output.toString('UTF-8').eachLine { line -> logger.error(line) }
-                execResult.assertNormalExitValue()
+            String refspec = System.getProperty("tests.bwc.refspec.${bwcBranch}", "${remote}/${bwcBranch}")
+            if (System.getProperty("tests.bwc.checkout.align") != null) {
+                /*
+                    We use a time based approach to make the bwc versions built deterministic and compatible with the current hash.
+                    Most of the time we want to test against latest, but when running delayed exhaustive tests or wanting
+                    reproducible builds we want this to be deterministic by using a hash that was the latest when the current
+                    commit was made.
+
+                    This approach doesn't work with merge commits as these can introduce commits in the chronological order
+                    after the fact e.x. a merge done today can add commits dated with yesterday so the result will no longer be
+                    deterministic.
+
+                    We don't use merge commits, but for additional safety we check that no such commits exist in the time period
+                    we are interested in.
+
+                    Timestamps are at seconds resolution. rev-parse --before and --after are inclusive w.r.t the second
+                    passed as input. This means the results might not be deterministic in the current second, but this
+                    should not matter in practice.
+                */
+                String timeOfCurrent = execGit { spec ->
+                    spec.commandLine 'git', 'show', '--no-patch', '--no-notes', "--pretty='%cD'"
+                    spec.workingDir project.rootDir
+                }
+                logger.lifecycle("Commit date of current: {}", timeOfCurrent)
+                String mergeCommits = execGit { spec ->
+                    spec.commandLine "git", "rev-list", refspec, "--after", timeOfCurrent, "--merges"
+                }
+                if (mergeCommits.isEmpty() == false) {
+                    throw new IllegalStateException(
```

Comment:
Ideally, what we would doo in this scenario is simply back up to the most recent commit before the merge but we can play it by ear. I suspect we won't encounter this too often, and if we do, we just have to wait until the next non-merge commit.

Reply:
Looking back over the last 6 months, I see 2 merge commits, one in particular in January was a manually merged feature branch. While we should not be doing that, we don't have any protections against them appear in the history right now, and failing the build on them existing would be an immediate breakage in CI that could not be fixed.

I suggested we use `--first-parent` in the rev-list command for traversing history. This will ensure we stay on the branch and don't fall into any commits that were made on a feature branch should another commit like that be added. We should then not need to check for merges.

Reply:
I don't have a string preference on the matter. 
We don't have any protections against it, but this could at least act as a check instead of making it be lenient. 
We check for merge commits only __after__ the time of the current commit, so creating an empty non merge commit after the merge commit would solve the problem of CI failing leaving the failed job to raise awareness.

Reply:
Merge commits for feature branches are acceptable, so that we don't lose history from long-running feature branches worked on by multiple developers (think: CCR, ILM, Zen 2).

Reply:
@jasontedor allowing them in this scenario could mean that we end up picking a commit form within the feature branch. If we just walk the first parent, we'll not have the feature branch in bwc but likely have it in master. In either case this will lead to a more obscure failure. I think I prefer the explicit failure here on the rare occasions, especially that all it takes is to have one commit after it, so one could push an empty commit after the merge commit ( with a lather date ) to avoid the fallout from CI